XenModel/Network/Heartbeat.cs

```diff
@@ -148,6 +148,23 @@ private void DoHeartbeat()
                 HandleConnectionLoss();
                 return;
             }
+            catch (WebException exn)
+            {
+                log.Error(exn);
+                if (((HttpWebResponse)exn.Response).StatusCode == HttpStatusCode.ProxyAuthenticationRequired) // work-around for CA-214653
+                {
+                    log.DebugFormat("Heartbeat for {0} has failed due to {1} credentials; closing the main connection",
+                        session == null ? "null" : session.Url,
+                        session.proxy.Proxy.Credentials == null ? "missing" : "incorrect");

```

Comment:
So the way I understand it then is that any formatted string ideally shouldn't have any argument which can only be of one language (e.g. {1} in this case, which can only be "missing" or "incorrect")? I will change that.

But arguments which are data (such as links, numbers, UUIDs, etc) are fine to keep (as they are localized 'by nature')? If the answer to this is yes then is the string "null" considered localised or not? Is it okay to keep line 207, for example, which uses the "null" string when session is null:
`log.DebugFormat("Heartbeat for {0} has failed; retrying", session == null ? "null" : session.Url);`
..or should these instances also be changed?